apps/omg_child_chain/lib/omg_child_chain/ethereum_event_aggregator.ex

```diff
@@ -0,0 +1,269 @@
+    empty_blknum_signature_events =
+      from_block..to_block
+      |> Enum.to_list()
+      |> Enum.map(fn blknum -> Enum.map(event_signatures, fn signature -> {blknum, signature, []} end) end)
+      |> List.flatten()
+
+    # we now merge the two lists
+    # it is important that logs we got from RPC are first
+    # because uniq_by takes the first occurance of {blknum, signature}
+    # so that we don't overwrite retrieved logs
+    data =
+      decoded_logs_in_keypair
+      |> Enum.concat(empty_blknum_signature_events)
+      |> Enum.uniq_by(fn {blknum, signature, _data} ->
+        {blknum, signature}
+      end)
+
+    :ets.insert(state.ets_bucket, data)
+  end
+
+  # delete everything older then (current block - delete_events_threshold)
+  defp delete_old_logs(new_height_blknum, state) do
+    # :ets.fun2ms(fn {block_number, _event_signature, _event} when
+    # block_number <= new_height - delete_events_threshold -> true end)
+    match_spec = [
+      {{:"$1", :"$2", :"$3"},
+       [{:"=<", :"$1", {:-, {:const, new_height_blknum}, {:const, state.delete_events_threshold_height_blknum}}}],
+       [true]}
+    ]
+
+    :ets.select_delete(state.ets_bucket, match_spec)
+  end
+
+  # allow ethereum event listeners to retrieve logs from ETS in bulk
+  defp handout_log(signature, from_block, to_block, state) do
+    # :ets.fun2ms(fn {block_number, event_signature, event} when
+    # block_number >= from_block and block_number <= to_block
+    # and event_signature == signature -> event and
+    # end)
+    event_match_spec = [
+      {{:"$1", :"$2", :"$3"},
+       [
+         {:andalso, {:andalso, {:>=, :"$1", {:const, from_block}}, {:"=<", :"$1", {:const, to_block}}},
+          {:==, :"$2", {:const, signature}}}
+       ], [:"$3"]}
+    ]
+
+    block_range = [
+      {{:"$1", :"$2", :"$3"},
+       [
+         {:andalso, {:andalso, {:>=, :"$1", {:const, from_block}}, {:"=<", :"$1", {:const, to_block}}},
+          {:==, :"$2", {:const, signature}}}
+       ], [:"$1"]}
+    ]
+
+    events = state.ets_bucket |> :ets.select(event_match_spec) |> List.flatten()
+    blknum_list = :ets.select(state.ets_bucket, block_range)
+
+    # we may not have all the block information the ethereum event listener wants
+    # so we check for that and find all logs for missing blocks
+    # in one RPC call for all signatures
+    case Enum.to_list(from_block..to_block) -- blknum_list do
+      [] ->
+        events
+
+      missing_blocks ->
+        missing_blocks = Enum.sort(missing_blocks)
+        missing_from_block = List.first(missing_blocks)
+        missing_to_block = List.last(missing_blocks)
+
+        _ =
+          Logger.debug(
+            "Missing block information (#{missing_from_block}, #{missing_to_block}) in event fetcher. Additional RPC call to gather logs."
+          )
+
+        true = retrieve_and_store_logs(missing_from_block, missing_to_block, state)
```

Comment:
yes it is! but `ets` is kinda unstandardised around returning values, so instead of rewriting return statements I'm asserting towards what :ets gives me back. https://erlang.org/doc/man/ets.html#insert-2

Reply:
Aaah I see, it does make sense. Although, I would argue that it might be nice to hide the idiosyncrasies of the tools we're relying on to offer a more uniform approach throughout our codebase. Just something to think about :)